### ISSUE TITLE
derivable_impls wants derive(Default) on enum

### ISSUE THREAD

**Issue by Arnavion** [rust-lang/rust-clippy#7753]
Lint name: [`derivable_impls`](https://rust-lang.github.io/rust-clippy/master/index.html#derivable_impls)

I tried [this code:](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=4a7402ede26296690167908770a84331)

```rust
pub enum IntOrString {
    Int(i32),
    String(String),
}

impl Default for IntOrString {
    fn default() -> Self {
        IntOrString::Int(0)
    }
}
```

Clippy says:

```
warning: this `impl` can be derived
  --> src/lib.rs:6:1
   |
6  | / impl Default for IntOrString {
7  | |     fn default() -> Self {
8  | |         IntOrString::Int(0)
9  | |     }
10 | | }
   | |_^
   |
   = note: `#[warn(clippy::derivable_impls)]` on by default
   = help: try annotating `IntOrString` with `#[derive(Default)]`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#derivable_impls
```

This is incorrect because rustc cannot derive `Default` on enums (until [`feature(derive_enum_default)`](https://github.com/rust-lang/rust/issues/87517) is stabilized.)

### Meta

**Rust version (`rustc -Vv`):**

```
rustc 1.57.0-nightly (2b862bed9 2021-09-23)
binary: rustc
commit-hash: 2b862bed9889808b69629fd7246317189b9517a5
commit-date: 2021-09-23
host: x86_64-unknown-linux-gnu
release: 1.57.0-nightly
LLVM version: 13.0.0
```

The playground repro uses `2021-10-02 f03eb6bef8ced8a24385`

---

cc @HKalbasi

@rustbot label +I-suggestion-causes-error

**Comment by HKalbasi** [comment]
🤦‍♂️

@rustbot claim

**Comment by AlexApps99** [comment]
I found a similar bug with unions, I assume this may fix that too.
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=617310465c57a719c089bb30374fb0f4

**Comment by AlexApps99** [comment]
Here's another one I found:
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=55789d00abff60aeba1fa92ff45de978
